latte/src/main/java/gg/beemo/latte/util/SuspendingRatelimit.kt

```diff
@@ -6,32 +6,56 @@ import kotlinx.coroutines.sync.withPermit
 import kotlin.time.Duration
 
 class SuspendingRatelimit(private val burst: Int, private val duration: Duration) {
-
     @Volatile
     private var remainingQuota: Int = burst
+
     @Volatile
     private var resetTimestamp: Long = 0
 
     private val quotaRequestSem = Semaphore(1)
 
-    fun overrideRatelimit(remainingQuota: Int, resetTimestamp: Long) {
+    fun overrideRatelimit(
+        remainingQuota: Int,
+        resetTimestamp: Long,
+    ) {
         this.remainingQuota = remainingQuota
         this.resetTimestamp = resetTimestamp
     }
 
+    private fun calculateWaitTime(): Long {
+        return (resetTimestamp - System.currentTimeMillis()).coerceAtLeast(0)
+    }
+
+    private fun tryResetQuota() {
+        if (System.currentTimeMillis() >= resetTimestamp) {
+            remainingQuota = burst
+            resetTimestamp = System.currentTimeMillis() + duration.inWholeMilliseconds
+        }
+    }
+
     suspend fun requestQuota() {
         quotaRequestSem.withPermit {
             if (remainingQuota <= 0) {
-                val waitTime = (resetTimestamp - System.currentTimeMillis()).coerceAtLeast(0)
+                val waitTime = calculateWaitTime()
                 delay(waitTime)
             }
-            if (System.currentTimeMillis() >= resetTimestamp) {
-                remainingQuota = burst
-                resetTimestamp = System.currentTimeMillis() + duration.inWholeMilliseconds
-            }
+            tryResetQuota()
+
             check(remainingQuota > 0)
             remainingQuota--
         }
     }
 
+    suspend fun tryRequestQuota(): Pair<Boolean, Long?> =
+        quotaRequestSem.withPermit {
```

Comment:
This suspends if another coroutine is currently waiting for a ratelimit permit in `requestQuota`, which is probably the one thing you wanted to avoid here 😄
Here, you can just take out the entire semaphore wrapper (and remove the `suspend`) and it should work as expected under concurrency.

Reply:
Hmm, so the semaphore is no longer in need for this case, am I reading it correctly? 💭 Was wondering if it would be safe to remove the semaphore for the "tryRequestQuota", but thought it wasn't, so I kept it there 😅

Reply:
Yeah, the way `requestQuota` works is that the `delay` blocks the semaphore (and thus other concurrent calls to the function) until new quota is available, which isn't needed here.